### DISCUSSION TITLE
Building C/C++ modules with micropython-wrap

### DISCUSSION THREAD

**Discussion by Tobinator32** [micropython/micropython#16260] [ESP32]
Hello folks,
i've been playing with micropython-wrap for some days now and am finally far enough that it compiles but it doesnt link. My example is based on the usercppmodule example micropython provides. Now i am getting multiple definition linker errors. In fact all functions inside my cpp files are duplicated. Does anyone has an idea how to troubleshoot or fix it?
Kind Regards Tobi

**Comment by bidrohini** [comment]
If you are including header files that define functions or variables, ensure that they are properly guarded to avoid multiple inclusions.
Use include guards:
#ifndef MY_HEADER_H
#define MY_HEADER_H



#endif // MY_HEADER_H

**Reply by Tobinator32** [reply]
Thanks for the reply.
i dont think this is the cause. All multiple definitions originate from the following:
```
#ifdef __cplusplus
extern "C"
{
#endif

using namespace upywrap;

struct F
{
    //func_name_def(cobsr_encode_buffer1)
    func_name_def(cobsr_decode_buffer1)
};

// std::string cobsr_encode_buffer1(std::string buffer_in)
// {
//     std::string data_in = buffer_in;
//     uint8_t buffer[512];
//     cobsr_encode_result res = cobsr_encode(buffer, sizeof(buffer), (const uint8_t*)(data_in.data()), data_in.size());

//     //py::bytes retval = py::bytes((const char*)buffer, res.out_len);
//     std::string retval ((const char*)buffer, res.out_len);
//     return retval;
// }

std::string cobsr_decode_buffer1(std::string buffer_in)
{
    std::string data_in = buffer_in;
    uint8_t buffer[512];
    cobsr_decode_result res = cobsr_decode(buffer, sizeof(buffer), (const uint8_t*)(data_in.data()), data_in.size());

    //py::bytes retval = py::bytes((const char*)buffer, res.out_len);
    std::string retval ((const char*)buffer, res.out_len);
    return retval;
}

    void doinit_upywraptest(mp_obj_dict_t* mod)
    {
        //Note this one must stay because this function can get called directly,
        //so without CreateModule() already calling InitializePyObjectStore().
        upywrap::InitializePyObjectStore( *mod );

        upywrap::FunctionWrapper fn(mod);
        fn.Def<F::cobsr_decode_buffer1>(cobsr_decode_buffer1);
        //fn.Def<F::cobsr_decode_buffer1>(cobsr_decode_buffer1);
    }

    mp_obj_module_t* init_upywraptest(void)
    {
        auto mod = upywrap::CreateModule( "upywraptest" );
        doinit_upywraptest( mod->globals );
        return mod;
    }

#ifdef __cplusplus
}
#endif
```
All functions here marked as duplicated.
These functions do not appear in any header

**Comment by stinos** [comment]
> All functions here marked as duplicated

Maybe they are duplicated? Build with V=1, carefully inspect the compilation phase that this file is really compiled only once, and check the linking phase that the output doesn't contain the .o twice. Or something like that.


> #ifdef __cplusplus
extern "C"
{
#endif

This isn't being used correctly here. It might work but it's confusing. This file is a .cpp, right, so contains C++ code so there's no way it can be compiled as C, hence the `#ifdef __cplusplus` isn't needed, it'll always be true. And the `extern "C"` (which is to remove C++ name mangling etc) should only apply to functions which get called directly from C, so only `init_upywraptest` I assume

**Reply by Tobinator32** [reply]
Yeah, maybe they are but im unable to find out why. Building a cpp example based on the provided one from micropython it works and runs fine. When im now including up-wrap they get duplicated. Will investigate further today. I think im very close to get it running. 
PS: i know how to use extern C etc. it was just an act of despair to get it running by accident. Ofcourse it makes very little sense but i think you get it ;)
Do you mind answering a few other questions that are offtopic and are not covered by the docu?

**Reply by Tobinator32** [reply]
Ok i think i tracked it down. When i gradually apply the neccessary up-wrap functions it begins throwing multiple definition errors when adding:`fn.Def<F::example_fun1>(example_fun1);` inside doinit_upywraptest (based on examples of up-wrap). Interestingly every function within this file is now duplicated although not touched by this.

**Reply by stinos** [reply]
Do you mean `example_fun1` gets a duplicate definition? Honestly no idea how that is possible; it's one file so if it's compiled only once I don't see how there can be multiple definitions. Does an `inline` for the function fix it? If so maybe somehow the .cpp is being included as a header somewhere? But that seems unlikely if micropython's own cppexample works. .

>  Do you mind answering a few other questions that are offtopic and are not covered by the docu?

Sure but maybe start a discussion at https://github.com/stinos/micropython-wrap/discussions then instead of here.

**Reply by Tobinator32** [reply]
No not only that but all functions from that translation unit/file.
See:
```
/opt/esp/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: esp-idf/main_esp32/libmain_esp32.a(example.cpp.obj): in function `StdString':
/project/libs/micropython/examples/usercmodule/cppexample/example.cpp:23: multiple definition of `StdString'; CMakeFiles/micropython.elf.dir/project/libs/micropython/examples/usercmodule/cppexample/example.cpp.obj:example.cpp:(.text+0x5c): first defined here
/opt/esp/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: esp-idf/main_esp32/libmain_esp32.a(example.cpp.obj): in function `cppfunc':
/project/libs/micropython/examples/usercmodule/cppexample/example.cpp:59: multiple definition of `cppfunc'; CMakeFiles/micropython.elf.dir/project/libs/micropython/examples/usercmodule/cppexample/example.cpp.obj:example.cpp:(.text+0xc8): first defined here
/opt/esp/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: esp-idf/main_esp32/libmain_esp32.a(example.cpp.obj): in function `cobsr_decode_buffer':  
/project/libs/micropython/examples/usercmodule/cppexample/example.cpp:79: multiple definition of `cobsr_decode_buffer'; CMakeFiles/micropython.elf.dir/project/libs/micropython/examples/usercmodule/cppexample/example.cpp.obj:example.cpp:(.text+0x1dc): first defined here
/opt/esp/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: esp-idf/main_esp32/libmain_esp32.a(example.cpp.obj):/project/libs/micropython-wrap/functionwrapper.h:129: multiple definition of `_ZN7upywrap15FunctionWrapper16functionPointersE'; CMakeFiles/micropython.elf.dir/project/libs/micropython/examples/usercmodule/cppexample/example.cpp.obj:(.bss+0x0): first defined here 
/opt/esp/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: esp-idf/main_esp32/libmain_esp32.a(example.cpp.obj): in function `doinit_upywraptest':   
/project/libs/micropython/examples/usercmodule/cppexample/example.cpp:39: multiple definition of `doinit_upywraptest'; CMakeFiles/micropython.elf.dir/project/libs/micropython/examples/usercmodule/cppexample/example.cpp.obj:example.cpp:(.text+0x74): first defined here
```
That are basically all function in this example. Removing `fn.Def` call results in a correct binary.
Maybe its worth mentioning, that there are two example.cpp.obj files in different directories. But why is this only a problem when introducing one function call?

**Reply by stinos** [reply]
Don't have time to do a complete write-up now, but strangely enough when I make a minimal sample: module.cpp with all test includes removed, and almost all code and all `Def` calls except the following removed

```
    upywrap::FunctionWrapper fn( mod );
    fn.Def< F::HasExceptions >( HasExceptions );
```

it actually does link. Note that `HasExceptions` is `inline`.

**Reply by Tobinator32** [reply]
do you mind sharing your makefile?

**Reply by stinos** [reply]
I couldn't get it to build anymore so probably made a mistake earlier. Anyway this makes it clear: https://github.com/micropython/micropython/issues/16284
